File: COMP602_Gymmate/app/src/main/java/com/example/gymmate/dialog/NumPickView.java
package com.example.gymmate.dialog;

import static java.lang.Math.abs;
import static java.lang.Math.min;

import android.animation.Animator;
import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.LinearInterpolator;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.example.gymmate.R;

import java.util.ArrayList;
import java.util.List;

public class NumPickView extends ScrollView {
    private String selectedTextColor="#FF0000";//被选中的文字的颜色
    private String unSelectedTextColor="#FFFFFF";//未被选中的文字的颜色
    private String dividerColor="#eeeeee";//分割线的颜色
    private int itemPadding=10;//item上下的padding，单位dp
    private int itemTextSize=16;//item的字体大小，单位sp
    public static final String TAG = NumPickView.class.getSimpleName();

    public interface OnWheelViewListener {
        void onSelected(int selectedIndex, String item);
    }

    private Context context;

    private LinearLayout views;

    public NumPickView(Context context) {
        super(context);
        init(context);
    }

    public NumPickView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public NumPickView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context);
    }

    List<String> items;

    private List<String> getItems() {
        return items;
    }

    public void setItems(List<String> list) {
        if (null == items) {
            items = new ArrayList<String>();
        }
        items.clear();
        items.addAll(list);

        // 前面和后面补全
        for (int i = 0; i < offset; i++) {
            items.add(0, "");
            items.add("");
        }

        initData();

    }


    public static final int OFF_SET_DEFAULT = 1;
    int offset = OFF_SET_DEFAULT; // 偏移量（需要在最前面和最后面补全）

    public int getOffset() {
        return offset;
    }

    public void setOffset(int offset) {
        this.offset = offset;
    }

    int displayItemCount; // 每页显示的数量

    int selectedIndex = 1;


    private void init(Context context) {
        this.context = context;
        //Log.d(TAG, "parent: " + this.getParent());
        this.setVerticalScrollBarEnabled(false);

        views = new LinearLayout(context);
        views.setOrientation(LinearLayout.VERTICAL);
        this.addView(views);

        scrollerTask = new Runnable() {

            public void run() {

                int newY = getScrollY();
                if (initialY - newY == 0) { // stopped
                    final int remainder = initialY % itemHeight;
                    final int divided = initialY / itemHeight;
                    if (remainder == 0) {
                        selectedIndex = divided + offset;

                        onSeletedCallBack();
                    } else {
                        if (remainder > itemHeight / 2) {
                            NumPickView.this.post(new Runnable() {
                                @Override
                                public void run() {
                                    NumPickView.this.smoothScrollTo(0, initialY - remainder + itemHeight);
                                    selectedIndex = divided + offset + 1;
                                    onSeletedCallBack();
                                }
                            });
                        } else {
                            NumPickView.this.post(new Runnable() {
                                @Override
                                public void run() {
                                    NumPickView.this.smoothScrollTo(0, initialY - remainder);
                                    selectedIndex = divided + offset;
                                    onSeletedCallBack();
                                }
                            });
                        }


                    }


                } else {
                    initialY = getScrollY();
                    NumPickView.this.postDelayed(scrollerTask, newCheck);
                }
            }
        };


    }

    int initialY;

    Runnable scrollerTask;
    int newCheck = 50;

    public void startScrollerTask() {

        initialY = getScrollY();
        this.postDelayed(scrollerTask, newCheck);
    }

    private void initData() {
        displayItemCount = offset * 2 + 1;

        for (String item : items) {
            views.addView(createView(item));
        }

        refreshItemView(0);
    }

    int itemHeight = 0;

    private TextView createView(String item) {
        TextView tv = new TextView(context);
        tv.setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        tv.setSingleLine(true);
        tv.setTextSize(TypedValue.COMPLEX_UNIT_SP, itemTextSize);
        tv.setText(item);
        tv.setGravity(Gravity.CENTER);
        int padding = dip2px(itemPadding);
        tv.setPadding(padding, padding, padding, padding);
        if (0 == itemHeight) {
            itemHeight = getViewMeasuredHeight(tv);
            //Log.d(TAG, "itemHeight: " + itemHeight);
            views.setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, itemHeight * displayItemCount));
            LinearLayout.LayoutParams lp = (LinearLayout.LayoutParams) this.getLayoutParams();
            this.setLayoutParams(new LinearLayout.LayoutParams(lp.width, itemHeight * displayItemCount));
        }
        return tv;
    }


    @Override
    protected void onScrollChanged(int l, int t, int oldl, int oldt) {
        super.onScrollChanged(l, t, oldl, oldt);
        refreshItemView(t);
        if (t > oldt) {//向下滚动
            scrollDirection = SCROLL_DIRECTION_DOWN;
        } else {//向上滚动
            scrollDirection = SCROLL_DIRECTION_UP;

        }


    }

    private void refreshItemView(int y) {
        int position = y / itemHeight + offset;
        int remainder = y % itemHeight;
        int divided = y / itemHeight;

        if (remainder == 0) {
            position = divided + offset;
        } else {
            if (remainder > itemHeight / 2) {
                position = divided + offset + 1;
            }
        }

        int childSize = views.getChildCount();
        for (int i = 0; i < childSize; i++) {
            TextView itemView = (TextView) views.getChildAt(i);
            if (null == itemView) {
                return;
            }
            if (position == i) {
                itemView.setTextColor(Color.parseColor(selectedTextColor));//设置被选中的item的字体颜色
            } else {
                itemView.setTextColor(Color.parseColor(unSelectedTextColor));
            }
        }
    }

    /**
     * 获取选中区域的边界
     */
    int[] selectedAreaBorder;

    private int[] obtainSelectedAreaBorder() {
        if (null == selectedAreaBorder) {
            selectedAreaBorder = new int[2];
            selectedAreaBorder[0] = itemHeight * offset;
            selectedAreaBorder[1] = itemHeight * (offset + 1);
        }
        return selectedAreaBorder;
    }


    private int scrollDirection = -1;
    private static final int SCROLL_DIRECTION_UP = 0;
    private static final int SCROLL_DIRECTION_DOWN = 1;

    Paint paint;
    int viewWidth;



    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        //Log.d(TAG, "w: " + w + ", h: " + h + ", oldw: " + oldw + ", oldh: " + oldh);
        viewWidth = w;
        setBackgroundDrawable(null);
    }

    /**
     * 选中回调
     */
    private void onSeletedCallBack() {
        if (null != onWheelViewListener) {
            try {
                onWheelViewListener.onSelected(selectedIndex, items.get(selectedIndex));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

    }

    public void setSeletion(int position) {
        final int p = position;
        selectedIndex = p + offset;
        this.post(new Runnable() {
            @Override
            public void run() {
                NumPickView.this.smoothScrollTo(0, p * itemHeight);
            }
        });

    }

    public String getSeletedItem() {
        return items.get(selectedIndex);
    }

    public int getSeletedIndex() {
        return selectedIndex - offset;
    }


    @Override
    public void fling(int velocityY) {
        super.fling(velocityY / 3);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        if (ev.getAction() == MotionEvent.ACTION_UP) {

            startScrollerTask();
        }
        return super.onTouchEvent(ev);
    }

    private OnWheelViewListener onWheelViewListener;

    public OnWheelViewListener getOnWheelViewListener() {
        return onWheelViewListener;
    }

    public void setOnWheelViewListener(OnWheelViewListener onWheelViewListener) {
        this.onWheelViewListener = onWheelViewListener;
    }

    private int dip2px(float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    private int getViewMeasuredHeight(View view) {
        int width = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        int expandSpec = View.MeasureSpec.makeMeasureSpec(Integer.MAX_VALUE >> 2, View.MeasureSpec.AT_MOST);
        view.measure(width, expandSpec);
        return view.getMeasuredHeight();
    }

    public String getSelectedTextColor() {
        return selectedTextColor;
    }

    public void setSelectedTextColor(String selectedTextColor) {
        this.selectedTextColor = selectedTextColor;
    }

    public String getUnSelectedTextColor() {
        return unSelectedTextColor;
    }

    public void setUnSelectedTextColor(String unSelectedTextColor) {
        this.unSelectedTextColor = unSelectedTextColor;
    }

    public String getDividerColor() {
        return dividerColor;
    }

    public void setDividerColor(String dividerColor) {
        this.dividerColor = dividerColor;
    }

    public int getItemPadding() {
        return itemPadding;
    }

    public void setItemPadding(int itemPadding) {
        this.itemPadding = itemPadding;
    }

    public int getItemTextSize() {
        return itemTextSize;
    }

    public void setItemTextSize(int itemTextSize) {
        this.itemTextSize = itemTextSize;
    }
}

File: COMP602_Gymmate/app/src/main/java/com/example/gymmate/dialog/NumPickView.kt
package com.example.gymmate.dialog

import java.lang.Math.abs
import java.lang.Math.min
import android.animation.Animator
import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.LinearInterpolator
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import com.example.gymmate.R
import java.util.ArrayList
import java.util.List

class NumPickView : ScrollView {
    var selectedTextColor = "#FF0000" //被选中的文字的颜色
    var unSelectedTextColor = "#FFFFFF" //未被选中的文字的颜色
    var dividerColor = "#eeeeee" //分割线的颜色
    var itemPadding = 10 //item上下的padding，单位dp
    var itemTextSize = 16 //item的字体大小，单位sp

    interface OnWheelViewListener {
        fun onSelected(selectedIndex: Int, item: String?)
    }

    private var context: Context? = null
    private var views: LinearLayout? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init(context)
    }

    var items: List<String>? = null
    private fun getItems(): List<String>? {
        return items
    }

    fun setItems(list: List<String?>?) {
        if (null == items) {
            items = ArrayList<String>()
        }
        items.clear()
        items.addAll(list)

        // 前面和后面补全
        for (i in 0 until offset) {
            items.add(0, "")
            items.add("")
        }
        initData()
    }

    var offset = OFF_SET_DEFAULT // 偏移量（需要在最前面和最后面补全）
    var displayItemCount = 0 // 每页显示的数量
    var selectedIndex = 1
    private fun init(context: Context) {
        this.context = context
        //Log.d(TAG, "parent: " + this.getParent());
        this.setVerticalScrollBarEnabled(false)
        views = LinearLayout(context)
        views.setOrientation(LinearLayout.VERTICAL)
        this.addView(views)
        scrollerTask = object : Runnable() {
            fun run() {
                val newY: Int = getScrollY()
                if (initialY - newY == 0) { // stopped
                    val remainder = initialY % itemHeight
                    val divided = initialY / itemHeight
                    if (remainder == 0) {
                        selectedIndex = divided + offset
                        onSeletedCallBack()
                    } else {
                        if (remainder > itemHeight / 2) {
                            this@NumPickView.post(object : Runnable() {
                                @Override
                                fun run() {
                                    this@NumPickView.smoothScrollTo(
                                        0,
                                        initialY - remainder + itemHeight
                                    )
                                    selectedIndex = divided + offset + 1
                                    onSeletedCallBack()
                                }
                            })
                        } else {
                            this@NumPickView.post(object : Runnable() {
                                @Override
                                fun run() {
                                    this@NumPickView.smoothScrollTo(0, initialY - remainder)
                                    selectedIndex = divided + offset
                                    onSeletedCallBack()
                                }
                            })
                        }
                    }
                } else {
                    initialY = getScrollY()
                    this@NumPickView.postDelayed(scrollerTask, newCheck)
                }
            }
        }
    }

    var initialY = 0
    var scrollerTask: Runnable? = null
    var newCheck = 50
    fun startScrollerTask() {
        initialY = getScrollY()
        this.postDelayed(scrollerTask, newCheck)
    }

    private fun initData() {
        displayItemCount = offset * 2 + 1
        for (item in items!!) {
            views.addView(createView(item))
        }
        refreshItemView(0)
    }

    var itemHeight = 0
    private fun createView(item: String): TextView {
        val tv = TextView(context)
        tv.setLayoutParams(
            LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        )
        tv.setSingleLine(true)
        tv.setTextSize(TypedValue.COMPLEX_UNIT_SP, itemTextSize)
        tv.setText(item)
        tv.setGravity(Gravity.CENTER)
        val padding = dip2px(itemPadding.toFloat())
        tv.setPadding(padding, padding, padding, padding)
        if (0 == itemHeight) {
            itemHeight = getViewMeasuredHeight(tv)
            //Log.d(TAG, "itemHeight: " + itemHeight);
            views.setLayoutParams(
                LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    itemHeight * displayItemCount
                )
            )
            val lp: LinearLayout.LayoutParams = this.getLayoutParams() as LinearLayout.LayoutParams
            this.setLayoutParams(LayoutParams(lp.width, itemHeight * displayItemCount))
        }
        return tv
    }

    @Override
    protected fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
        super.onScrollChanged(l, t, oldl, oldt)
        refreshItemView(t)
        scrollDirection = if (t > oldt) { //向下滚动
            SCROLL_DIRECTION_DOWN
        } else { //向上滚动
            SCROLL_DIRECTION_UP
        }
    }

    private fun refreshItemView(y: Int) {
        var position = y / itemHeight + offset
        val remainder = y % itemHeight
        val divided = y / itemHeight
        if (remainder == 0) {
            position = divided + offset
        } else {
            if (remainder > itemHeight / 2) {
                position = divided + offset + 1
            }
        }
        val childSize: Int = views.getChildCount()
        for (i in 0 until childSize) {
            val itemView: TextView = views.getChildAt(i) as TextView ?: return
            if (position == i) {
                itemView.setTextColor(Color.parseColor(selectedTextColor)) //设置被选中的item的字体颜色
            } else {
                itemView.setTextColor(Color.parseColor(unSelectedTextColor))
            }
        }
    }

    /**
     * 获取选中区域的边界
     */
    var selectedAreaBorder: IntArray?
    private fun obtainSelectedAreaBorder(): IntArray {
        if (null == selectedAreaBorder) {
            selectedAreaBorder = IntArray(2)
            selectedAreaBorder!![0] = itemHeight * offset
            selectedAreaBorder!![1] = itemHeight * (offset + 1)
        }
        return selectedAreaBorder
    }

    private var scrollDirection = -1
    var paint: Paint? = null
    var viewWidth = 0
    @Override
    protected fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        //Log.d(TAG, "w: " + w + ", h: " + h + ", oldw: " + oldw + ", oldh: " + oldh);
        viewWidth = w
        setBackgroundDrawable(null)
    }

    /**
     * 选中回调
     */
    private fun onSeletedCallBack() {
        if (null != onWheelViewListener) {
            try {
                onWheelViewListener!!.onSelected(selectedIndex, items!![selectedIndex])
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun setSeletion(position: Int) {
        selectedIndex = position + offset
        this.post(object : Runnable() {
            @Override
            fun run() {
                this@NumPickView.smoothScrollTo(0, position * itemHeight)
            }
        })
    }

    val seletedItem: String
        get() = items!![selectedIndex]
    val seletedIndex: Int
        get() = selectedIndex - offset

    @Override
    fun fling(velocityY: Int) {
        super.fling(velocityY / 3)
    }

    @Override
    fun onTouchEvent(ev: MotionEvent): Boolean {
        if (ev.getAction() === MotionEvent.ACTION_UP) {
            startScrollerTask()
        }
        return super.onTouchEvent(ev)
    }

    var onWheelViewListener: OnWheelViewListener? = null

    private fun dip2px(dpValue: Float): Int {
        val scale: Float = context.getResources().getDisplayMetrics().density
        return (dpValue * scale + 0.5f).toInt()
    }

    private fun getViewMeasuredHeight(view: View): Int {
        val width: Int = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        val expandSpec: Int =
            View.MeasureSpec.makeMeasureSpec(Integer.MAX_VALUE shr 2, View.MeasureSpec.AT_MOST)
        view.measure(width, expandSpec)
        return view.getMeasuredHeight()
    }

    companion object {
        val TAG: String = NumPickView::class.java.getSimpleName()
        const val OFF_SET_DEFAULT = 1
        private const val SCROLL_DIRECTION_UP = 0
        private const val SCROLL_DIRECTION_DOWN = 1
    }
}